First try

DemichevD opened the pull request:
Add classes: BooleanFieldFormat, FieldValidator,TableValidator, FieldFormatFactory

Add interface: Validator
Add tests: FieldFormatTest, TableFormatTest

DemichevD committed:
Add classes: BooleanFieldFormat, FieldValidator,TableValidator, FieldFormatFactory
Add interface: Validator
Add tests: FieldFormatTest, TableFormatTest

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldFormat.java:
```diff
@@ -40,18 +45,55 @@ public void setNullable(Boolean nullable) {
+        try {
+            return (FieldFormat<?>) super.clone();
+        } catch (CloneNotSupportedException e) {
+            throw new AssertionError("Невозможно");
```
лучше new IllegalStateException(e). не нужно проглатывать эксепшен которые прилетает в катче. Его лучше прокидывать дальше.

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldFormatFactory.java:
```diff
@@ -2,14 +2,84 @@
-        ff.setName(name);
-        ff.setDescription(description);
-        return ff;
+    public static FieldFormat<?> createFieldFormat(Character type, String name, String description, Boolean nullable, Boolean hidden, String defaultValue) {
```
Дубляж функций, надо отрефакторить так, чтобы не повторялся код

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -0,0 +1,54 @@
+
+    private final ArrayList<String> errorMessage = new ArrayList<>();
+
+    public void Validation(FieldFormat<?> value) {
```
методы класса всегда идут в camelCase и не должны начинаться с большой буквы

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -0,0 +1,54 @@
+    public void Validation(FieldFormat<?> value) {
+        FieldValidator fieldValidator = new FieldValidator();
+        if (fieldValidator.isValid(value)) {
+            System.out.println("Ok");
```
вместо вывода в консоль, надо бросать эксепшен что данные не валидные

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -0,0 +1,54 @@
+
+import java.util.ArrayList;
+
+public class FieldValidator implements Validator<FieldFormat<?>> {
```
Валидатор не должен знать про филдформат и не должен его принимать.

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Override
+    public boolean isValid(FieldFormat<?> value) {
```
функция должна принимать какое то значение ячейки и проверять валидно оно или нет. но не должно принимать филдформат

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -0,0 +1,54 @@
+
+
+    @Override
+    public void getErrorMessage() {
```
надо бросать эксепшен

AndreyPlis commented on src/main/java/com/tibbo/datatable/TableFormat.java:
```diff
@@ -3,16 +3,79 @@
+            for (FieldFormat<?> f : fields)
+                copy.fields.add(f.clone());
+        } catch (CloneNotSupportedException e) {
+            throw new AssertionError("Невозможно");
```
лучше IllegalStateException(e) плюс сообщение. не проглатывай эксепшен в катче.

AndreyPlis commented on src/test/java/FieldFormatTest.java:
```diff
@@ -1,27 +1,119 @@
 
     @Test
-    public void createStringFieldFormatFromFactory() {
+    public void checkConvertToString() {
```
Между тестами не должно шариться переменных или состояний. Для каждого теста поднимается свое окружение и свои переменные.Значит филдформаты в каждом тесте свои и заново создаются

AndreyPlis commented on src/test/java/FieldFormatTest.java:
```diff
@@ -1,27 +1,119 @@
+        System.out.println("All tests finished");
+    }
+
+    public static void main(String[] args) {
```
мейн не нужен в тестах

AndreyPlis commented on src/test/java/FieldFormatTest.java:
```diff
@@ -1,27 +1,119 @@
+
+}
+
+class CalcListenerFieldFormat extends RunListener {
```
не нужен

AndreyPlis commented on src/test/java/TableFormatTest.java:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+class CalcListenerTableFormat extends RunListener {
```
аналогично

AndreyPlis commented on src/main/java/com/tibbo/datatable/TableValidator.java:
```diff
@@ -0,0 +1,42 @@
+    private final ArrayList<String> errorMessage = new ArrayList<>();
+
+    public void Validation(TableFormat value) {
+        TableValidator tableValidator = new TableValidator();
```
зачем внутри обычного метода создавать новый обьект валидатора?. Учитывая что функция не статическая. значит мне сначала создать обьект валидатор, потом вызвать функцию validation в которой валидатор по новой создаться.

AndreyPlis commented on src/main/java/com/tibbo/datatable/TableFormat.java:
```diff
@@ -3,16 +3,79 @@
+    public TableFormat() {
+    }
+
+    public TableFormat(int min, int max) {
```
не хватает проверки на то что мин меньше макса

DemichevD committed:
Fixed TableFormatTest, FieldFormatTest: Each test creates its own variables;
Fixed FieldFormat: corrected throw new exception in clone method;
Fixed TableFormat: corrected throw new exception in clone method, added validation of input parameters in the TableFormat constructor;
Fixed FieldFormatFactory: removed unnecessary class constructors;

DemichevD committed:
Fixed FieldFormatFactory: add function "funcToCreateFieldFormat"

DemichevD committed:
Add abstract class Validators
Fixed interface Validator
Fixed methods in FieldValidator

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -1,54 +1,37 @@
+import java.util.regex.Pattern;
 
-public class FieldValidator implements Validator<FieldFormat<?>> {
+public class FieldValidator extends Validators {
```
должен быть не  один клас который делает проверки, а по классу на каждую проверку

AndreyPlis commented on src/main/java/Main.java:
```diff
@@ -1,14 +1,20 @@
-        fieldFormat.setName("One");
-        fieldFormat.setHidden(false);
-        fieldValidator.Validation(fieldFormat);
+        FieldFormat<?> fieldFormat = FieldFormatFactory.createFieldFormat(FieldFormat.STRING_FIELD,"Alex",null,null,null,"One");
```
такие нуллы  не очень при вызове). это значит что не совсем удобно создавать обьект

AndreyPlis commented on src/main/java/Main.java:
```diff
@@ -1,14 +1,20 @@
+        FieldFormat<?> fieldFormat = FieldFormatFactory.createFieldFormat(FieldFormat.STRING_FIELD,"Alex",null,null,null,"One");
+        TableFormat tableFormat = new TableFormat(15,25);
+        String number = null;
+        //nullFieldValidator(fieldFormat.getName());
```
комментарии удаляй если не нужны

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -1,54 +1,37 @@
+    @Override
+    public <T> void nullFieldValidator(T value) {
+        if (value == null)
+            throw new NullPointerException("Значение не может быть равно нулю");
```
старайся в коде писать только на английском

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldValidator.java:
```diff
@@ -1,54 +1,37 @@
-        for (String s : errorMessage)
-            System.out.println(s);
+    public void lengthFieldValidator(String value) {
+        int lengthField = 50;
```
почему именно 50?

AndreyPlis commented on src/main/java/com/tibbo/datatable/TableValidator.java:
```diff
@@ -2,21 +2,40 @@
 
     @Override
-    public boolean isValid(TableFormat value) {
+    public  void regexValidation(String value) {
```
пустые методы, это следствие неккоректного наследования. Филд валидатор был расчитан на филды. Валидаторы на весь тайбл формат очевидно отдельный

AndreyPlis commented on src/main/java/com/tibbo/datatable/Validators.java:
```diff
@@ -0,0 +1,16 @@
+package com.tibbo.datatable;
+
+public abstract class Validators implements Validator{
```
Почему множесмтвенно число? Валидаторы наследуются от одного валидатора? это крайне странно. Это как комплект резины наследуется от одного колеса.

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldFormatFactory.java:
```diff
@@ -3,82 +3,56 @@
+        return fieldFormat;
+    }
+    public static FieldFormat<?> createFieldFormat(Character type, String name, String description, Boolean nullable, Boolean hidden) {
+        FieldFormat<?> fieldFormat = funcToCreateFieldFormat(type);
```
все еще дубляж. ты делаешь одни и те же действия в разных функциях

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldFormat.java:
```diff
@@ -92,7 +92,7 @@ public String toString() {
             return (FieldFormat<?>) super.clone();
         } catch (CloneNotSupportedException e) {
-            throw new AssertionError("Невозможно");
+            throw new IllegalStateException("Невозможно", e);
```
английский. и сообщение должно быть понятным. Что значит невозможно?)) невозможно что?)

DemichevD committed:
Class Validators rename in FieldValidator(abstract class);
TableValidator rename in LimitsValidator;
Add classes: LengthFieldValidator, NullFieldValidator, RegexValidator;
Fixed FieldFormatFactory.

AndreyPlis commented on src/main/java/com/tibbo/datatable/FieldFormatFactory.java:
```diff
@@ -55,4 +19,20 @@
+
+    @SafeVarargs
+    private static <T> void setArguments(FieldFormat<?> fieldFormat, T... args) {
+        switch (args.length) {
```
не надежный код. а что если я при создании формата просто поменяю порядок аргументов?

DemichevD committed:
add class DataRecord
add method validate in FieldFormat

DemichevD committed:
Fixed method valueFromString in BooleanFieldFormat, IntFieldFormat;
TableFormat - Rename method getFields in getField;
DataRecord - rename method addDataToList  - addValue, rename method changeDataToList  - setValue, add method getValue;
